extensions/amp-accordion/1.0/accordion.js

```diff
@@ -0,0 +1,176 @@
+
+  const registerSection = useCallback((id) => {
+    return () => setExpandedSet((expandedSet) => omit(expandedSet, id));
+  }, []);
+
+  const toggleExpanded = useCallback(
+    (id, defaultExpanded) => {
+      setExpandedSet((expandedSet) => {
+        const newValue = !isExpanded(
+          id,
+          defaultExpanded,
+          expandedSet,
+          expandSingleSection
+        );
+        let newExpandedSet;
+        if (newValue && expandSingleSection) {
+          newExpandedSet = {[id]: newValue};
+          for (const k in expandedSet) {
+            if (k != id) {
+              newExpandedSet[k] = false;
+            }
+          }
+        } else {
+          newExpandedSet = {...expandedSet, [id]: newValue};
+        }
+        return newExpandedSet;
+      });
+    },
+    [expandSingleSection]
+  );
+
+  const context = useMemo(
+    () =>
+      /** @type {!AccordionDef.ContextProps} */ ({
+        registerSection,
+        toggleExpanded,
+        isExpanded: (id, defaultExpanded) =>
+          isExpanded(id, defaultExpanded, expandedSet, expandSingleSection),
+      }),
+    [expandedSet, expandSingleSection, registerSection, toggleExpanded]
+  );
+
+  return (
+    <Comp {...rest}>
+      <AccordionContext.Provider value={context}>
+        {children}
+      </AccordionContext.Provider>
+    </Comp>
+  );
+}
+
+/**
+ * @param {string} id
+ * @param {boolean} defaultExpanded
+ * @param {!Object<string, boolean>} expandedSet
+ * @param {boolean} expandSingleSection
+ * @return {boolean}
+ */
+function isExpanded(id, defaultExpanded, expandedSet, expandSingleSection) {
+  const current = expandedSet[id];
+  if (current != null) {
+    return current;
+  }
+  const adjDefaultExpanded = expandSingleSection
+    ? Object.values(expandedSet).includes(true)
+      ? false
+      : defaultExpanded
+    : defaultExpanded;
+  return adjDefaultExpanded;
+}
+
+/**
+ * @param {!AccordionDef.SectionProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function AccordionSection({
+  as: Comp = 'section',
+  headerAs: HeaderComp = 'header',
+  contentAs: ContentComp = 'div',
+  expanded: defaultExpanded = false,
+  header,
+  children,
+  ...rest
+}) {
+  const [id] = useState(generateSectionId);
+  const [expandedState, setExpandedState] = useState(defaultExpanded);
+
+  const {registerSection, isExpanded, toggleExpanded} = useContext(
+    AccordionContext
+  );
+
+  useEffect(() => registerSection && registerSection(id), [
+    registerSection,
+    id,
+  ]);
+
+  const expandHandler = useCallback(() => {
+    if (toggleExpanded) {
+      toggleExpanded(id, defaultExpanded);
+    } else {
+      setExpandedState((prev) => !prev);
+    }
+  }, [id, toggleExpanded, defaultExpanded]);
+
+  const expanded = isExpanded ? isExpanded(id, defaultExpanded) : expandedState;
+
+  return (
+    <Comp {...rest} expanded={expanded} aria-expanded={String(expanded)}>
+      <HeaderComp role="button" onClick={expandHandler}>
```

Comment:
An interesting a11y consideration is that buttons should be clickable via `Space` or `Enter` -- we don't have to act on this now but the browser `<button>` element supports `Space`/`Enter` => `click()` whereas the `role="button"` does not. 

We could reverse defaults for `HeaderComp` to tag `"button"` with `role="heading"` instead, but it's not too critical either way as we can always follow up to add the listeners ourselves.

There are a few other aria labels that [accordions are expected to have](https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html), but these can be added separately if preferred.

Reply:
Separately, while this is `role="button"`, please also add `tabindex="0"`

Reply:
For sure. I was going to add trackers for a11y and do a separate "porting" pass. But I wanted to get the outline merged first to be able to split the remainder of work a bit more.